src/main/java/org/codehaus/groovy/ast/expr/RangeExpression.java

```diff
@@ -27,26 +27,42 @@
  * <pre>for i in 0..10 {...}</pre>
  */
 public class RangeExpression extends Expression {
-
     private final Expression from;
     private final Expression to;
-    private final boolean inclusive;
+    private final boolean inclusive; // Kept to keep old code depending on this working
+    // GROOVY-9649
+    private final boolean exclusiveLeft;
+    private final boolean exclusiveRight;
 
+    // Kept until sure this can be removed
     public RangeExpression(Expression from, Expression to, boolean inclusive) {
         this.from = from;
         this.to = to;
         this.inclusive = inclusive;
+        this.exclusiveLeft = false;
+        this.exclusiveRight = !inclusive;
+        setType(ClassHelper.RANGE_TYPE);
```

Comment:
Can this constructor delegate to the new constructor?

Reply:
Agreed, I'll make this change.